### PR TITLE
Arachni plugin

### PR DESCRIPTION
**Pull request by brandonprry**
This plugin adds support for driving an Arachni scan and saving the results in the Metasploit database. However, it seems that web vulns aren't displayed in the ```vulns``` table in msfconsole. This might be a bug, I'm not sure, but I vaguely remember having the same issue when developing the sqlmap plugin.

```
msf > load arachni
[*] Arachni plugin loaded.
[*] Successfully loaded plugin: Arachni
msf > arachni_connect
msf > arachni_scan http://demo.testfire.net/default.aspx "xss*"
msf > arachni_scanlog
[+] 1. Cross-Site Scripting (XSS)
[+] Scan running: true
msf > arachni_scanlog
[+] 1. Cross-Site Scripting (XSS)
[+] 2. Cross-Site Scripting (XSS) in HTML tag
[+] 3. Cross-Site Scripting (XSS)
[+] Scan running: false
msf > arachni_savelog
msf > vulns
msf > web_vulns
[-] Unknown command: web_vulns.
msf > 
```

I am probably going to make arachni_scanlog print a Rex::Table, but I wanted to ask about web vulns being displayed in the ```vulns``` table.

### CODE REVIEW

**Comment by brandonprry** [comment]
Oh, The arachni classes in Rex::Proto were shamelessly stolen from the arachni-rpc-pure Github repo, but these are license with 2- or 3-clause BSD, which I think is compatible with Metasploit.

File: lib/rex/proto/arachni/connection.rb [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+
+            context.key  =
+                OpenSSL::PKey::RSA.new( File.open( options[:ssl_pkey] ) )
+
+            context.ca_file     = options[:ssl_ca]
+            context.verify_mode =
+                OpenSSL::SSL::VERIFY_PEER | OpenSSL::SSL::VERIFY_FAIL_IF_NO_PEER_CERT
+        end
+
+        @socket = OpenSSL::SSL::SSLSocket.new(
```
Comment (sempervictus) [inline lib/rex/proto/arachni/connection.rb]: Would probably make sense to use Rex::Socket::TcpSsl for this to allow us to access Arachni installs on the other side of a compromised host (for instance if deploying the scanner as a form of payload for rapid internal web scans of the environment).
Forcing TLS validation may also be a problem in some cases, though optional validation is definitely a good thing (even a good default, just suggesting the option of NO_VERIFY).
[Rex Socket SSL client certificate support](https://github.com/rapid7/rex-socket/pull/6) may help a bit for this.

File: lib/rex/proto/arachni/connection.rb [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+        receive_object
+    end
+
+    def send_object( obj )
+        serialized = serialize( obj )
+        @socket.puts( [ serialized.bytesize, serialized ].pack( 'Na*' ) )
+    end
+
+    def receive_object
+        while data = @socket.sysread( 99999 )
```
Comment (sempervictus) [inline lib/rex/proto/arachni/connection.rb]: May want to comment this process for intent and functionality - what's being recv'd, unpacked, and unserialized.

File: lib/rex/proto/arachni/connection.rb [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    # @return   [String]
+    #   Decompressed string.
+    def try_decompress( string )
+        # Just an ID representing a serialized, empty data structure.
+        return string if string.size == 1
+
+        begin
+            Zlib::Inflate.inflate string
+        rescue Zlib::DataError
+            string
```
Comment (sempervictus) [inline lib/rex/proto/arachni/connection.rb]: Any danger in potentially returning incompletely received binary data here?

**Review by sempervictus — COMMENTED** [review]
Awesome! Thank you.
Big fan of the proper separation of plugin CRUD and library components for interface. Would even suggest pushing a few more pieces to the client such as the dispatcher and instance init.

Going to try and put it through its paces tomorrow, run some functional tests and see how it works with scan grids.

Only real issue i see on first pass is the sockets use here, admittedly as i'm on a crusade to eliminate all ...Socket.new calls in the codebase (pivot all the things!). 

Thanks again

File: plugins/arachni.rb [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    def commands
+      {
+        'arachni_connect' => 'Connect to an Arachni RPC instance',
+        'arachni_scan' => 'Scan a URL',
+        'arachni_scanlog' => 'Get the log for the scan',
+        'arachni_savelog' => 'Save the results of the scan to the database'
+      }
+    end
+
+    def cmd_arachni_connect(*args)
```
Comment (sempervictus) [inline plugins/arachni.rb]: If the localhost isnt the Arachni master, this approach blows up:
```
[-] Error while running command arachni_connect: Connection refused - connect(2) for "127.0.0.1" port 7331

Call stack:
/opt/metasploit4/msf4/lib/rex/proto/arachni/connection.rb:51:in `initialize'
/opt/metasploit4/msf4/lib/rex/proto/arachni/connection.rb:51:in `new'
/opt/metasploit4/msf4/lib/rex/proto/arachni/connection.rb:51:in `initialize'
/opt/metasploit4/msf4/lib/rex/proto/arachni/client.rb:62:in `new'
/opt/metasploit4/msf4/lib/rex/proto/arachni/client.rb:62:in `with_connection'
/opt/metasploit4/msf4/lib/rex/proto/arachni/client.rb:53:in `call'
/opt/metasploit4/msf4/plugins/arachni.rb:28:in `cmd_arachni_connect'
```

File: plugins/arachni.rb [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+      @dispatcher = Rex::Proto::Arachni::Client.new(
+        host: args[0] || '127.0.0.1',
+        port: args[1] ||7331
+      )
+
+      instance_info = @dispatcher.call('dispatcher.dispatch', Rex::Text.rand_text_alpha(8))
+      @instance = Rex::Proto::Arachni::Client.new(
```
Comment (sempervictus) [inline plugins/arachni.rb]: See above regarding remote masters

File: lib/rex/proto/arachni/connection.rb (hunk 3 above)
Comment (brandonprry) [inline lib/rex/proto/arachni/connection.rb]: Potentially, I suppose, but I'm not sure of a better way to handle the situation. This is functionally equivalent of checking the first few header bytes to determine if the string is actually zlib-compressed, and, if not, just return the data as is.

Comment (brandonprry) [inline lib/rex/proto/arachni/connection.rb]: You can't know if the data is zlib compressed without testing it.

**Comment by brandonprry** [comment]
This won't work well with scan grids. I don't provide much power in managing multiple instances. You can't kick off multiple scans and track them, only at a time. I wanted to keep it simple and easy to use.

**Comment by pbarry-r7** [comment]
Hey @bperry-r7, you might try loading the wmap plugin (`load wmap`) and see if the web vulns show up with its `wmap_vulns -l` command:

```
msf > load wmap

.-.-.-..-.-.-..---..---.
| | | || | | || | || |-'
`-----'`-'-'-'`-^-'`-'
[WMAP 1.5.1] ===  et [  ] metasploit.com 2012
[*] Successfully loaded plugin: wmap
msf > wmap_vulns
[*] Usage: wmap_vulns [options]
	-h 		Display this help text
	-l 		Display web vulns table

msf > wmap_vulns -l
[*] + [10.0.0.1] (mysite.info): imported //xmlrpc.php
[*] 	The GHOST Vulnerability 
[*] 	POST <empty response>
[*] + [10.0.0.1] (mysite.info): imported /
[*] 	WordPress Cross-Site Scripting Vulnerability 
[*] 	GET <empty response>
[*] + [10.0.0.1] (mysite.info): imported /
[*] 	WordPress Multiple Cross-Site Scripting Vulnerabilities 
[*] 	GET <empty response>
.
.
.
```

**Comment by brandonprry** [comment]
You're right, that works to display the vulns in the database. Thanks! They should also appear in the community/pro interface regardless (as long as they are in the same workspace).

**Comment by brandonprry** [comment]
Unfortunately I've not had time to wrap this up. I will soon though.

**Comment by acammack-r7** [comment]
Hey @brandonprry, in order to help us have a better handle on which PRs are currently in progress, I am closing this for now and adding the `attic` label. You can learn out our new PR label [on our wiki](https://github.com/rapid7/metasploit-framework/wiki/So-Your-PR-was-closed#moved-to-the-attic). Feel free to reopen once work resumes, I love having integrations to other tools!

**Comment by bcoles** [comment]
Arachni was [archived](https://github.com/Arachni/arachni) in January 2020 and is [no longer supported](https://www.arachni-scanner.com/blog/arachni-is-no-longer-maintained/).

**Pull request closed without merging**